File: FunctionalPatterns/Behavioral/Strategy.fs
namespace FunctionalPatterns
namespace FunctionalPatterns.Behavioral

module Strategy =

    type SpeakingType =
        { speak: string -> unit }

    let screaming =
        { speak = fun (text: string) -> printfn "%s" (text.ToUpper()) }
    let slowTalking =
        { speak = fun (text: string) -> printfn "%s" (System.String.Join(' ', text.ToCharArray())) }

    let speakText speakingType text =
        speakingType.speak text

File: ObjectOrientedPatterns/Behavioral/Strategy.cs
namespace ObjectOrientedPatterns.Behavioral.Strategy
{
    public interface ISpeakingStrategy
    {
        public void Speak(string text);
    }
    public class Screaming : ISpeakingStrategy
    {
        public void Speak(string text) => Console.WriteLine(text.ToUpper());
    }
    public class SlowTalk : ISpeakingStrategy
    {
        public void Speak(string text) => Console.WriteLine(string.Join(' ', text.ToCharArray()));
    }

    public class Speaking
    {
        public Speaking(ISpeakingStrategy strategy)
        {
            myStrategy = strategy;
        }
        public void SpeakText(string text)
        {
            myStrategy.Speak(text);
        }
        private ISpeakingStrategy myStrategy;
    }
}
